File: 90.subsets-ii.java
/*
 * @lc app=leetcode id=90 lang=java
 *
 * [90] Subsets II
 */

// @lc code=start
class Solution {
    public List<List<Integer>> subsetsWithDup(int[] nums) {
        List<List<Integer>> rtn = new ArrayList<>();
        Arrays.sort(nums);
        gen(rtn, new ArrayList<Integer>(), nums, 0);
        return rtn;
    }

    public void gen(List<List<Integer>> rtn, ArrayList<Integer> cur, int[] nums, int idx) {
        rtn.add(new ArrayList<Integer>(cur));
        for (int i = idx; i < nums.length; i++) {
            if (i > idx && nums[i - 1] == nums[i])
                continue;
            cur.add((Integer) nums[i]);
            gen(rtn, cur, nums, i + 1);
            cur.remove((Integer) nums[i]);
        }
    }
}
// @lc code=end

File: 90.subsets-ii.py
#
# @lc app=leetcode id=90 lang=python3
#
# [90] Subsets II
#
# https://leetcode.com/problems/subsets-ii/description/
#
# algorithms
# Medium (55.57%)
# Likes:    7388
# Dislikes: 207
# Total Accepted:    651.8K
# Total Submissions: 1.2M
# Testcase Example:  '[1,2,2]'
#
# Given an integer array nums that may contain duplicates, return all possible
# subsets (the power set).
#
# The solution set must not contain duplicate subsets. Return the solution in
# any order.
#
#
# Example 1:
# Input: nums = [1,2,2]
# Output: [[],[1],[1,2],[1,2,2],[2],[2,2]]
# Example 2:
# Input: nums = [0]
# Output: [[],[0]]
#
#
# Constraints:
#
#
# 1 <= nums.length <= 10
# -10 <= nums[i] <= 10
#
#
#

# @lc code=start
class Solution:
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        l = len(nums)
        rtn = []

        def dfs(path=[], idx=0):
            rtn.append(path[:])
            for i in range(idx, l):
                cur = nums[i]
                if i > idx and cur == nums[i - 1]:
                    continue
                path.append(cur)
                dfs(path, i + 1)
                path.pop()

        dfs()
        return rtn


# @lc code=end
